lightning/src/ln/peers/handshake/states.rs

```diff
@@ -0,0 +1,906 @@
+	}
+	pub(super) fn new_responder(responder_static_private_key: &SecretKey, responder_ephemeral_private_key: &SecretKey) -> Self {
+		HandshakeState::ResponderAwaitingActOne(ResponderAwaitingActOneState::new(responder_static_private_key.clone(), responder_ephemeral_private_key.clone()))
+	}
+}
+
+impl IHandshakeState for HandshakeState {
+	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String> {
+		match self {
+			HandshakeState::InitiatorStarting(state) => { state.next(input) },
+			HandshakeState::ResponderAwaitingActOne(state) => { state.next(input) },
+			HandshakeState::InitiatorAwaitingActTwo(state) => { state.next(input) },
+			HandshakeState::ResponderAwaitingActThree(state) => { state.next(input) },
+			HandshakeState::Complete(_conduit) => { panic!("no acts left to process") }
+		}
+	}
+}
+
+// Handshake state of the Initiator prior to generating Act 1
+pub(super) struct InitiatorStartingState {
+	initiator_static_private_key: SecretKey,
+	initiator_static_public_key: PublicKey,
+	initiator_ephemeral_private_key: SecretKey,
+	initiator_ephemeral_public_key: PublicKey,
+	responder_static_public_key: PublicKey,
+	chaining_key: Sha256,
+	hash: Sha256
+}
+
+// Handshake state of the Responder prior to receiving Act 1
+pub(super) struct ResponderAwaitingActOneState {
+	responder_static_private_key: SecretKey,
+	responder_ephemeral_private_key: SecretKey,
+	responder_ephemeral_public_key: PublicKey,
+	chaining_key: Sha256,
+	hash: Sha256,
+	act_one_builder: ActBuilder
+}
+
+// Handshake state of the Initiator prior to receiving Act 2
+pub(super) struct InitiatorAwaitingActTwoState {
+	initiator_static_private_key: SecretKey,
+	initiator_static_public_key: PublicKey,
+	initiator_ephemeral_private_key: SecretKey,
+	responder_static_public_key: PublicKey,
+	chaining_key: ChainingKey,
+	hash: Sha256,
+	act_two_builder: ActBuilder
+}
+
+// Handshake state of the Responder prior to receiving Act 3
+pub(super) struct ResponderAwaitingActThreeState {
+	hash: Sha256,
+	responder_ephemeral_private_key: SecretKey,
+	chaining_key: ChainingKey,
+	temporary_key: [u8; 32],
+	act_three_builder: ActBuilder
+}
+
+impl InitiatorStartingState {
+	pub(crate) fn new(initiator_static_private_key: SecretKey, initiator_ephemeral_private_key: SecretKey, responder_static_public_key: PublicKey) -> Self {
+		let initiator_static_public_key = private_key_to_public_key(&initiator_static_private_key);
+		let (hash, chaining_key) = initialize_handshake_state(&responder_static_public_key);
+		let initiator_ephemeral_public_key = private_key_to_public_key(&initiator_ephemeral_private_key);
+		InitiatorStartingState {
+			initiator_static_private_key,
+			initiator_static_public_key,
+			initiator_ephemeral_private_key,
+			initiator_ephemeral_public_key,
+			responder_static_public_key,
+			chaining_key,
+			hash
+		}
+	}
+}
+
+impl IHandshakeState for InitiatorStartingState {
```

Comment:
AFAICT, the trait is never really used, even in tests - here and for all the individual messages, we could just have a "dumb" impl block and things would work as well, and then the trait is only ever used by the HandshakeState enum, which means it could just be dropped.

Reply:
It is a defensive programming technique to ensure all of the individual state enums all implement the same common interface. That way, the wrapper enum can be simple delegation.

It was the idiomatic rust enum dispatch method I found. Other projects leverage macros and crates, but I chose to keep it more verbose to not introduce additional dependencies.

Here is the trait definition that I hope explains it. If you want this removed just let me know, I don't need to bikeshed about it.

```
// Trait for all individual states to implement that ensure HandshakeState::next() can
// delegate to a common function signature. May transition to the same state in the event there are
// not yet enough bytes to move forward with the handshake.
pub(super) trait IHandshakeState {
	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String>;
}
```

Reply:
Hmm, I guess I don't have a super strong feeling on using a trait to show identical operation, though it wasn't immediately obvious to me that that was its purpose given its public across the module (but only used in the states.rs file) and implemented by both the top-level enum and structs inside. Maybe best to move it in states.rs and add a comment?

Reply:
I'm not sure I see the value of `IHandshakeState`. Nothing enforces that the enum states must implement this trait.